### PR TITLE
Introduce Health. Checks for Ditto Missing Permissions

### PR DESCRIPTION
**Pull request by anyercastillo**
Ticket: https://github.com/getditto/DittoAndroidTools/issues/22
This PR partially implements the ticket.
- permissions are granted for Ditto,

Other acceptances criteria will be implemented in future PRs.
- Bluetooth is turned on, and
- LAN is turned on.

### Video 
https://github.com/getditto/DittoAndroidTools/assets/5747647/710f5156-1097-4626-baf7-b57410192efa

### CODE REVIEW

File: DittoHealth/src/main/res/xml/provider_paths.xml
Comment (byoung-ditto) [inline DittoHealth/src/main/res/xml/provider_paths.xml:1]: Maybe not necessary, perhaps a copy/paste leftover

File: app/src/main/java/live/ditto/dittotoolsapp/MainActivity.kt [hunk 1]
```diff
@@ -37,7 +38,7 @@ class MainActivity : ComponentActivity() {
         lateinit var ditto: Ditto
         try {
             val androidDependencies = DefaultAndroidDittoDependencies(applicationContext)
-            val identity = DittoIdentity.OnlinePlayground(androidDependencies, appId = "YOUR_APPID", token = "YOUR_TOKEN", enableDittoCloudSync = true)
+            val identity = DittoIdentity.OnlinePlayground(androidDependencies, appId = "c5adf242-f3fc-4987-aac9-3cb79967274e", token = "812872a2-28a5-4ee7-b487-04f452eb6176", enableDittoCloudSync = true)
```
Comment (byoung-ditto) [inline app/src/main/java/live/ditto/dittotoolsapp/MainActivity.kt:41]: Is the AppId/token addition intentional?

File: DittoHealth/src/test/java/live/ditto/dittoexportlogs/ExampleUnitTest.kt
Comment (byoung-ditto) [inline DittoHealth/src/test/java/live/ditto/dittoexportlogs/ExampleUnitTest.kt:1]: The  package path dittoexportlogs may be a copy/paste leftover

File: DittoHealth/src/main/java/live/ditto/health/usecase/GetDittoMissingPermissionsFlow.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import live.ditto.transports.DittoSyncPermissions
+import kotlin.coroutines.coroutineContext
+
+class GetDittoMissingPermissionsFlow(
+    private val context: Context,
+) {
+    operator fun invoke(
+        timeoutMs: Long = 300
+    ) = flow {
+        while (coroutineContext.isActive) {
```
Comment (byoung-ditto) [inline DittoHealth/src/main/java/live/ditto/health/usecase/GetDittoMissingPermissionsFlow.kt:16]: Looks good, wonder if in future we can do this without looping? Perhaps a permissions listener or Android Broadcast Receiver of some sort

File: DittoHealth/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
+
+<!--    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>-->
+
+
+    <application>
+        <provider
```
Comment (byoung-ditto) [inline DittoHealth/src/main/AndroidManifest.xml:8]: Maybe not necessary, perhaps a copy/paste leftover

File: DittoHealth/src/androidTest/java/live/ditto/dittoexportlogs/ExampleInstrumentedTest.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+@RunWith(AndroidJUnit4::class)
+class ExampleInstrumentedTest {
+    @Test
+    fun useAppContext() {
+        // Context of the app under test.
+        val appContext = InstrumentationRegistry.getInstrumentation().targetContext
+        assertEquals("live.ditto.dittoexportlogs.test", appContext.packageName)
```
Comment (byoung-ditto) [inline DittoHealth/src/androidTest/java/live/ditto/dittoexportlogs/ExampleInstrumentedTest.kt:22]: Maybe wrong context/path? perhaps a copy/paste leftover

**Comment by byoung-ditto** [comment]
When testing I was warned of 1 or 4 permissions in different cases

Install app, upon first launch click "Dont Allow" when prompted to allow permissions
	The app displays Red/Not Healthy with 4 missing permissions

In contrast install app, upon first launch click "Allow" when prompted to allow permissions
	The app displays Green/Healthy
	Now go into Android App settings, and manually remove the "Nearby devices" permission for "DittoToolsApp"
	Reopen the app, it displayed Red/Not Healthy with 1 mission permission

Havent investigated why it displays different list of permissions (1 vs 4) in these two cases, but could this being confusing for developers as well as app end users

Screenshot_20240129_194043
Screenshot_20240129_194314

**Comment by byoung-ditto** [comment]
A couple final notes:

1) when "Nearby devices" is disallowed in the Android app settings, when opening the DittoToolsApp, a white screen is displayed for a few seconds. Recommend we verify this behavior is an artifact of the test app, and not of the "Ditto Health" module that will be embedded in customer apps

2) In some cases, I was prompted to allow Android Notifications. Even when tapping "Dont Allow" the Notifications permission was not listed in the Health Viewer, and did not seem to impact the Red vs Green status.

3) The UI/view is great for developers and tech support. If this is intended for end users, maybe consider adding a more human readable messages like those displaed in the Android App settings e.g. "Nearby devices" rather than "android.permission.NEARBY_WIFI_DEVICES" could be the main/large label, with the full permission string as a sublabel

File: DittoHealth/src/main/java/live/ditto/health/usecase/GetDittoMissingPermissionsFlow.kt (hunk 2 above)
Comment (anyercastillo) [reply DittoHealth/src/main/java/live/ditto/health/usecase/GetDittoMissingPermissionsFlow.kt:16]: Permissions are not observables in Android.
I am afraid is not other way to do it :(

File: DittoHealth/src/main/AndroidManifest.xml (hunk 3 above)
Comment (anyercastillo) [reply DittoHealth/src/main/AndroidManifest.xml:8]: Yeah,
I copy/paste one of the modules and I did not pay attention to the manifest. 
My bad.

File: DittoHealth/src/androidTest/java/live/ditto/dittoexportlogs/ExampleInstrumentedTest.kt (hunk 4 above)
Comment (anyercastillo) [reply DittoHealth/src/androidTest/java/live/ditto/dittoexportlogs/ExampleInstrumentedTest.kt:22]: These tests have no values, I will delete the files.
thanks

File: app/src/main/java/live/ditto/dittotoolsapp/MainActivity.kt (hunk 1 above)
Comment (anyercastillo) [reply app/src/main/java/live/ditto/dittotoolsapp/MainActivity.kt:41]: Hi, 
Good catch,
It is not intentional, I change those value to test the app. 
It should not be part of the PR. 
I will remove it.

Thanks

**Comment by anyercastillo** [comment]
Thanks for the test!
Let me address some of your comments:

The module always prints the missing permissions returned by the Ditto SDK, it is hard for me to explain why the function returns 1 vs 4 permissions, I think is worth to take a look to the function implementation to understand the reason.

The white screen is the result of the MainScreen [instantiating Ditto on the Main thread](https://github.com/getditto/DittoAndroidTools/blob/99eadb2667e9f5782fc1f4c4d1e1e80ebe917cf7/app/src/main/java/live/ditto/dittotoolsapp/MainActivity.kt#L42). Please notice it is not possible for the component to render a white screen, once the composable is loaded by the navigation controller, it will render a green or red screen.
```
@Composable
private fun HealthScreen(state: HealthUiState) {
    if (state.isHealthy) {
        Healthy()
    } else {
        NotHealthy(state)
    }
}
```

File: DittoHealth/src/main/java/live/ditto/health/HealthScreen.kt [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            )
+
+        Spacer(modifier = Modifier.height(16.dp))
+
+        state.causes.forEach { cause ->
+            Cause(cause = cause)
+            Spacer(modifier = Modifier.height(16.dp))
+        }
+    }
+}
```
Comment (bplattenburg) [inline DittoHealth/src/main/java/live/ditto/health/HealthScreen.kt:91]: Minor - I'm not sure how much effort its worth putting into a quick demo app like this, but does it make sense to usehealthy / unhealthy state from the `HealthUiState` here directly and collapse `Healthy()` and `NotHealthy(state:)` into a single composable that swaps the colors based on the health check? It seems like maybe there's some duplicated UI logic that could be simplified

File: DittoHealth/src/test/java/live/ditto/health/ExampleUnitTest.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
+    @Test
+    fun addition_isCorrect() {
+        assertEquals(4, 2 + 2)
+    }
+}
```
Comment (bplattenburg) [inline DittoHealth/src/test/java/live/ditto/health/ExampleUnitTest.kt:17]: I think we can remove the sample test

**Review by bplattenburg — COMMENTED** [review]
Looks great! I left a few minor comments but nothing blocking

File: DittoHealth/src/main/java/live/ditto/health/HealthScreen.kt (hunk 5 above)
Comment (anyercastillo) [reply DittoHealth/src/main/java/live/ditto/health/HealthScreen.kt:91]: I am not following you, 
will ping you in Slack, so we can talk more about it.

Comment (bplattenburg) [reply DittoHealth/src/main/java/live/ditto/health/HealthScreen.kt:91]: After our slack conversation, there's not really any duplication going on here, and I need to get another coffee.

Comment (anyercastillo) [reply DittoHealth/src/main/java/live/ditto/health/HealthScreen.kt:91]: @bplattenburg and I met on Slack and we agreed on the following.
- Move the logic to resolve texts from the composable to the UiState model
- Keep the existing structure of Healthy/NotHealthy composables

**Comment by okdistribute** [comment]
Thanks Anyer!

Can we focus the language to be slightly more for an end-user? For example, 

* instead of "system is healthy" say "all permissions granted" / hard to know what system you're referring to
* When missing location permissions, is it possible to show a button that when pressed will force the operating system ask for permissions again?

File: DittoHealth/src/test/java/live/ditto/health/ExampleUnitTest.kt (hunk 6 above)
Comment (anyercastillo) [reply DittoHealth/src/test/java/live/ditto/health/ExampleUnitTest.kt:17]: no op, for now.
until we figure out the path for other modules too.
if we decide to remove tests, then we should remove in all modules

**Comment by anyercastillo** [comment]
Sure thing Rae

**Comment by anyercastillo** [comment]
With the [last commit](https://github.com/getditto/DittoAndroidTools/pull/29/commits/40287b22988cd9230ae78dfb43f5d911dfacdd66) this is how looks the UI

https://github.com/getditto/DittoAndroidTools/assets/5747647/35e903f2-86ac-47d9-b8d4-f58bd8c79b9b

**Comment by anyercastillo** [comment]
Let's wait for https://github.com/getditto/DittoAndroidTools/pull/30
Then I will rebase and fix any conflict

**Comment by anyercastillo** [comment]
Note: The latest main contains the required code to publish artifacts. 
After rebasing on main, let's add a commit on top to address deployment.

**Comment by okdistribute** [comment]
LGTM

**Review by marinakuksoditto — APPROVED** [review]
Looks great!

**Pull request merged**